web/interceptor/access_control.go

```diff
@@ -0,0 +1,157 @@
+)
+
+type (
+	role      int
+	roles     []role
+	requestID interface {
+		IDFor(string) uint64
+	}
+)
+
+const (
+	none role = iota
+	// user role implies that user attempts to access information about himself.
+	user
+	// group role implies that the user is a course student + a member of the given group.
+	group
+	// student role implies that the user is enrolled in the course with any role.
+	student
+	// teacher: user enrolled in the course with teacher status.
+	teacher
+	// courseAdmin: an admin user who is also enrolled into the course.
+	courseAdmin
+	// admin is the user with admin privileges.
+	admin
+)
+
+// If there are several roles that can call a method, a role with the least privilege must come first.
+// If method is not in the map, there is no restrictions to call it.
+var access = map[string]roles{
+	"GetUser":                {none},
+	"GetCourse":              {none},
+	"GetCourses":             {none},
+	"CreateEnrollment":       {user},
+	"UpdateCourseVisibility": {user},
+	"GetCoursesByUser":       {user},
+	// TODO(vera): needs a specific check: if request attempts to change admin role, user role is not sufficien.
+	"UpdateUser":              {user, admin},
+	"GetEnrollmentsByUser":    {user, admin},
+	"GetSubmissions":          {group, student, teacher, courseAdmin},
+	"GetGroupByUserAndCourse": {group, teacher},
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetAssignments":          {student, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"GetRepositories":         {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {teacher, courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		logger.Debugf("ACCESS CONTROL for method %s", method) // tmp
+		roles, ok := access[method]
+		if ok {
```

Comment:
This should be changed to:
```go
if !ok {
  // error logging and return early...
}
// Then here comes the ok part (avoiding the indentation).
```
This because it is difficult to read the code, when you have to scroll to the bottom to check that the erroneous case is handled appropriately. Is it?